sources/healthdog/src/service_check.rs

```diff
@@ -0,0 +1,127 @@
+use crate::error::{self, Result};
+use lazy_static::lazy_static;
+use log::trace;
+use regex::Regex;
+use snafu::ResultExt;
+use std::process::Command;
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) struct ServiceHealth {
+    /// Whether or not the service is healthy.
+    pub(crate) is_healthy: bool,
+    /// In the event of an unhealthy service, the service's exit code (if found).
+    pub(crate) exit_code: Option<i32>,
+}
+
+pub(crate) trait ServiceCheck {
+    /// Checks the given service to see if it is healthy.
+    fn check(&self, service_name: &str) -> Result<ServiceHealth>;
+}
+
+pub(crate) struct SystemdCheck {}
+
+impl ServiceCheck for SystemdCheck {
+    fn check(&self, service_name: &str) -> Result<ServiceHealth> {
+        if is_ok(service_name)? {
+            return Ok(ServiceHealth {
+                is_healthy: true,
+                exit_code: None,
+            });
+        }
+        Ok(ServiceHealth {
+            is_healthy: false,
+            exit_code: parse_service_exit_code(service_name)?,
+        })
+    }
+}
+
+struct Outcome {
+    exit: i32,
+    stdout: String,
+}
+
+impl Outcome {
+    fn is_exit_true(&self) -> bool {
+        self.exit == 0
+    }
+}
+
+fn systemctl(args: &[&str]) -> Result<Outcome> {
+    trace!("calling systemctl with '{:?}'", args);
+    let output = Command::new("systemctl")
+        .args(args)
+        .output()
+        .context(error::Command {
+            command: "systemctl",
+            args: args.iter().map(|&s| s.to_owned()).collect::<Vec<String>>(),
+        })?;
+    Ok(Outcome {
+        exit: output.status.code().unwrap_or(-1),
+        stdout: String::from_utf8_lossy(output.stdout.as_slice()).into(),
+    })
+}
+
+fn is_active(service: &str) -> Result<bool> {
+    let outcome = systemctl(&["is-active", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_failed(service: &str) -> Result<bool> {
+    let outcome = systemctl(&["is-failed", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_ok(service: &str) -> Result<bool> {
+    Ok(!is_failed(service)? && is_active(service)?)
+}
+
+fn parse_service_exit_code(service: &str) -> Result<Option<i32>> {
+    // we don't check the command's exit code because systemctl returns non-zero codes for various
+    // non-exceptional execution outcomes.
+    let outcome = systemctl(&["--no-pager", "status", service])?;
+    Ok(parse_stdout(&outcome.stdout)?)
+}
+
+/// Regex pattern for finding the exit code of a systemd service that has exited. There is a single
+/// capture group, named `exit_code`.
+const SYSTEMD_EXIT_REGEX_PATTERN: &str =
+    r#"Main PID: \d+ \(code=[a-zA-Z0-9-_]+, status=(?P<exit_code>\d{1,3})/[a-zA-Z0-9-_]+\)"#;
```

Comment:
I think the "/" part after the code is optional; at least, I'm not seeing it in an example on my machine.

In any case, I'd really like to avoid parsing this if we can.  `systemctl show` has a much simpler output format than `systemctl status`; it seems to be intended for machines.  There's a line like `ExecMainStatus=17` that would, obviously, be easier to parse; I wouldn't even use a regex.  Would that work for us?

Reply:
No longer applicable due to [11d0d7b](https://github.com/bottlerocket-os/bottlerocket/compare/cbc5d1d0b587442093072c22e97e1a9398d9f153..11d0d7bdd62c0b6555729a5854732b295227af9e)